File: app/src/main/java/ca/mudar/rotationquicksetting/ui/SettingsFragment.java
package ca.mudar.rotationquicksetting.ui;

import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceGroup;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.view.View;

import java.util.Locale;

import ca.mudar.rotationquicksetting.BuildConfig;
import ca.mudar.rotationquicksetting.Const;
import ca.mudar.rotationquicksetting.Const.PrefsNames;
import ca.mudar.rotationquicksetting.R;
import ca.mudar.rotationquicksetting.data.UserPrefs;
import ca.mudar.rotationquicksetting.utils.PermissionUtils;

import static ca.mudar.rotationquicksetting.utils.OrientationUtils.ROTATION_LAND;
import static ca.mudar.rotationquicksetting.utils.OrientationUtils.ROTATION_LAND_REVERSE;
import static ca.mudar.rotationquicksetting.utils.OrientationUtils.ROTATION_PORT;
import static ca.mudar.rotationquicksetting.utils.OrientationUtils.ROTATION_PORT_REVERSE;

/**
 * Created by mudar on 10/05/17.
 */

public class SettingsFragment extends PreferenceFragment implements
        SharedPreferences.OnSharedPreferenceChangeListener,
        Preference.OnPreferenceClickListener {

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    private SharedPreferences mSharedPrefs;
    private Preference mPermissionGranted;
    private SettingsAboutCallback mListener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof SettingsAboutCallback) {
            mListener = (SettingsAboutCallback) context;
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);


        final PreferenceManager pm = this.getPreferenceManager();
        pm.setSharedPreferencesName(Const.APP_PREFS_NAME);
        pm.setSharedPreferencesMode(Context.MODE_PRIVATE);

        addPreferencesFromResource(R.xml.prefs_settings);
        mPermissionGranted = findPreference(PrefsNames.PERMISSION_GRANTED);

        mSharedPrefs = pm.getSharedPreferences();

        mSharedPrefs.registerOnSharedPreferenceChangeListener(this);

        mPermissionGranted.setOnPreferenceClickListener(this);
        findPreference(PrefsNames.HELP).setOnPreferenceClickListener(this);
        findPreference(PrefsNames.ABOUT).setOnPreferenceClickListener(this);
        findPreference(PrefsNames.EULA).setOnPreferenceClickListener(this);

        hideHelpIfPossible();
    }

    @Override
    public void onResume() {
        super.onResume();

        setupSummaries();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        /**
         * Remove the listener
         */
        if (mSharedPrefs != null) {
            mSharedPrefs.unregisterOnSharedPreferenceChangeListener(this);
        }
    }

    /**
     * Implements SharedPreferences.OnSharedPreferenceChangeListener
     *
     * @param sharedPreferences
     * @param key
     */
    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (PrefsNames.REVERSE_PORT.equals(key)) {
            showTryRotationSnackbar(sharedPreferences.getBoolean(key, false) ?
                    ROTATION_PORT_REVERSE : ROTATION_PORT);
        } else if (PrefsNames.REVERSE_LAND.equals(key)) {
            showTryRotationSnackbar(sharedPreferences.getBoolean(key, false) ?
                    ROTATION_LAND_REVERSE : ROTATION_LAND);
        }
    }

    /**
     * Implements Preference.OnPreferenceClickListener
     *
     * @param preference
     * @return
     */
    @Override
    public boolean onPreferenceClick(Preference preference) {
        final String key = preference.getKey();
        if (PrefsNames.PERMISSION_GRANTED.equals(key)) {
            startActivity(PermissionUtils.getPermissionIntent(getContext()));
            return true;
        } else if (PrefsNames.HELP.equals(key)) {
            showHelp();
            return true;
        } else if (mListener != null) {
            if (PrefsNames.ABOUT.equals(key)) {
                mListener.onShowAbout();
                return true;
            } else if (PrefsNames.EULA.equals(key)) {
                mListener.onShowEula();
                return true;
            }
        }

        return false;
    }

    private void setupSummaries() {
        mPermissionGranted.setSummary(getPermissionGrantedSummary());

        findPreference(PrefsNames.ABOUT).setSummary(getString(
                R.string.prefs_about_summary, BuildConfig.VERSION_NAME));
    }

    @StringRes
    private int getPermissionGrantedSummary() {
        return Settings.System.canWrite(getContext()) ?
                R.string.prefs_permission_granted_on : R.string.prefs_permission_granted_off;
    }

    private void setOrientation(int orientation) {
        if (!Settings.System.canWrite(getContext())) {
            return;
        }

        final ContentResolver contentResolver = getContext().getContentResolver();
        Settings.System.putInt(contentResolver, Settings.System.ACCELEROMETER_ROTATION, 0);
        Settings.System.putInt(contentResolver, Settings.System.USER_ROTATION, orientation);
    }

    private void showTryRotationSnackbar(final int orientation) {
        if (getView() == null || !Settings.System.canWrite(getContext())) {
            return;
        }

        @StringRes int message;
        switch (orientation) {
            case ROTATION_PORT:
                message = R.string.snackbar_try_rotation_port;
                break;
            case ROTATION_PORT_REVERSE:
                message = R.string.snackbar_try_rotation_port_reverse;
                break;
            case ROTATION_LAND:
                message = R.string.snackbar_try_rotation_land;
                break;
            case ROTATION_LAND_REVERSE:
                message = R.string.snackbar_try_rotation_land_reverse;
                break;
            default:
                return;
        }
        Snackbar.make(getView(), message, Snackbar.LENGTH_LONG)
                .setAction(R.string.btn_try_it, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        setOrientation(orientation);
                    }
                })
                .setActionTextColor(ContextCompat.getColor(getContext(), R.color.snackbar_action_text))
                .show();
    }

    private void showHelp() {
        final String language = Locale.getDefault().getLanguage();
        final Intent viewIntent = new Intent(Intent.ACTION_VIEW);
        viewIntent.setData(Uri.parse(getResources()
                .getString(R.string.url_help_quick_settings, language)));
        startActivity(viewIntent);
    }

    private void hideHelpIfPossible() {
        final UserPrefs prefs = UserPrefs.getInstance(getContext());
        if (!prefs.hasHelp()) {
            ((PreferenceGroup) findPreference(PrefsNames.CAT_GENERAL))
                    .removePreference(findPreference(PrefsNames.HELP));
            prefs.setHelpCompleted();
        }
    }

    public interface SettingsAboutCallback {
        void onShowAbout();

        void onShowEula();
    }
}

File: app/src/main/java/ca/mudar/rotationquicksetting/ui/SettingsFragment.kt
/*
 * Rotation Quick Setting
 * Add a Quick Settings tile to select portrait or landscape screen orientation.
 *
 * Copyright (C) 2017 Mudar Noufal <mn@mudar.ca>
 *
 * This file is part of RotationQS.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ca.mudar.rotationquicksetting.ui

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.preference.Preference
import android.preference.PreferenceFragment
import android.preference.PreferenceGroup
import android.provider.Settings
import android.support.annotation.StringRes
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import ca.mudar.rotationquicksetting.BuildConfig
import ca.mudar.rotationquicksetting.Const
import ca.mudar.rotationquicksetting.Const.PrefsNames
import ca.mudar.rotationquicksetting.R
import ca.mudar.rotationquicksetting.data.UserPrefs
import ca.mudar.rotationquicksetting.utils.OrientationUtils.*
import ca.mudar.rotationquicksetting.utils.PermissionUtils
import java.util.*

class SettingsFragment : PreferenceFragment(),
        SharedPreferences.OnSharedPreferenceChangeListener,
        Preference.OnPreferenceClickListener {

    companion object {
        fun newInstance(): SettingsFragment {
            return SettingsFragment()
        }
    }

    private var mPermissionGranted: Preference? = null
    private var mListener: SettingsAboutCallback? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)

        if (context is SettingsAboutCallback) {
            mListener = context
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        preferenceManager.sharedPreferencesName = Const.APP_PREFS_NAME
        preferenceManager.sharedPreferencesMode = Context.MODE_PRIVATE

        addPreferencesFromResource(R.xml.prefs_settings)
        mPermissionGranted = findPreference(PrefsNames.PERMISSION_GRANTED)

        preferenceManager?.sharedPreferences?.registerOnSharedPreferenceChangeListener(this)

        mPermissionGranted?.onPreferenceClickListener = this

        findPreference(PrefsNames.HELP)?.onPreferenceClickListener = this
        findPreference(PrefsNames.ABOUT)?.onPreferenceClickListener = this

        hideHelpIfPossible()
    }

    override fun onResume() {
        super.onResume()

        setupSummaries()
    }

    override fun onDestroy() {
        super.onDestroy()

        /**
         * Remove the listener
         */
        preferenceManager?.sharedPreferences?.unregisterOnSharedPreferenceChangeListener(this)
    }

    /**
     * Implements SharedPreferences.OnSharedPreferenceChangeListener
     *
     * @param sharedPreferences
     * @param key
     */
    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (PrefsNames.REVERSE_PORT == key) {
            showTryRotationSnackbar(if (sharedPreferences.getBoolean(key, false))
                ROTATION_PORT_REVERSE
            else
                ROTATION_PORT)
        } else if (PrefsNames.REVERSE_LAND == key) {
            showTryRotationSnackbar(if (sharedPreferences.getBoolean(key, false))
                ROTATION_LAND_REVERSE
            else
                ROTATION_LAND)
        }
    }

    /**
     * Implements Preference.OnPreferenceClickListener
     *
     * @param preference
     * @return
     */
    override fun onPreferenceClick(preference: Preference): Boolean {
        val key = preference.key
        if (PrefsNames.PERMISSION_GRANTED == key) {
            startActivity(PermissionUtils.getPermissionIntent(context))
            return true
        } else if (PrefsNames.HELP == key) {
            showHelp()
            return true
        } else if (PrefsNames.ABOUT == key && mListener != null) {
            mListener!!.onShowAbout()
            return true
        }

        return false
    }

    private fun setupSummaries() {
        mPermissionGranted?.setSummary(getPermissionGrantedSummary())

        findPreference(PrefsNames.ABOUT)?.summary = getString(R.string.prefs_about_summary,
                BuildConfig.VERSION_NAME)
    }

    @StringRes
    private fun getPermissionGrantedSummary(): Int {
        return when (Settings.System.canWrite(context)) {
            true -> R.string.prefs_permission_granted_on
            false -> R.string.prefs_permission_granted_off
        }
    }

    private fun setOrientation(orientation: Int) {
        if (!Settings.System.canWrite(context)) {
            return
        }

        Settings.System.putInt(context.contentResolver, Settings.System.ACCELEROMETER_ROTATION, 0)
        Settings.System.putInt(context.contentResolver, Settings.System.USER_ROTATION, orientation)
    }

    private fun showTryRotationSnackbar(orientation: Int) {
        if (!Settings.System.canWrite(context)) {
            return
        }

        val message = when (orientation) {
            ROTATION_PORT -> R.string.snackbar_try_rotation_port
            ROTATION_PORT_REVERSE -> R.string.snackbar_try_rotation_port_reverse
            ROTATION_LAND -> R.string.snackbar_try_rotation_land
            ROTATION_LAND_REVERSE -> R.string.snackbar_try_rotation_land_reverse
            else -> return
        }

        view?.let {
            Snackbar.make(view, message, Snackbar.LENGTH_LONG)
                    .setAction(R.string.btn_try_it) { setOrientation(orientation) }
                    .setActionTextColor(ContextCompat.getColor(context, R.color.snackbar_action_text))
                    .show()
        }
    }

    private fun showHelp() {
        val language = Locale.getDefault().language
        val viewIntent = Intent(Intent.ACTION_VIEW)
        viewIntent.data = Uri.parse(resources
                .getString(R.string.url_help_quick_settings, language))
        startActivity(viewIntent)
    }

    private fun hideHelpIfPossible() {
        val prefs = UserPrefs(context)
        if (!prefs.hasHelp()) {
            (findPreference(PrefsNames.CAT_GENERAL) as PreferenceGroup)
                    .removePreference(findPreference(PrefsNames.HELP))
            prefs.setHelpCompleted()
        }
    }

    interface SettingsAboutCallback {
        fun onShowAbout()
    }
}
